tests/test_project/tests/test_qa.py

```diff
@@ -189,5 +207,23 @@ def test_qa_call_check_commit_message_failure(self):
                 else:
                     self.fail('SystemExit or Exception not raised')
 
+    def test_commit_failure_message(self):
+        bad_commit = [
+            'commitcheck',
+            '--message',
+            "[qa] Updated file and fixed problem\n\n"
+            "Added more files. Fixes #20"
+        ]
+        captured_output = io.StringIO()
+        sys.stdout = captured_output  # redirect stdout
+        with patch('argparse._sys.argv', bad_commit):
+            try:
+                check_commit_message()
+            except (SystemExit):
```

Comment:
```
except (SystemExit):
   pass
```
makes me a little nervous, if something fails, one cannot tell the reason without editing the tests.

Can you please do it like:
```
                except (SystemExit, Exception) as e:
                    msg = 'Check failed:\n\n{}' \
                          '\n\nOutput:{}'.format(option[-1], e)
                    self.fail(msg)
                else:
                     sys.stdout = sys.__stdout__  # reset redirect
                     message = 'Your commit message does not follow our commit message style guidelines'
                     self.assertIn(message, captured_output.getvalue())
```

Reply:
`SystemExit` in this test is supposed to be raised (look at these lines: https://github.com/openwisp/openwisp-utils/blob/master/openwisp_utils/qa.py#L157-#L159) and therefore if it will be raised everything in `else` clausule won't be executed.

We might supply user with message if `SystemExit` won't be raised (Exception or nothing would be raised), but wouldn't be better to do it in [this function](https://github.com/openwisp/openwisp-utils/blob/master/tests/test_project/tests/test_qa.py#L117)?

Reply:
Hey, 

Ah, I haven't checked it that closely, I only wanted to point out the risk of 
```
except:
   pass
```
and we should avoid as much as we can unless it's the only clean way to test!

If it's suppose to execute the `except` part, even better, can we do:
```
except(SystemExit):
    sys.stdout = sys.__stdout__  # reset redirect
    message = 'Your commit message does not follow our commit message style guidelines'
    self.assertIn(message, captured_output.getvalue()
else: 
    self.fail("Error not raised")
```
After all, we only want to test in case the "except" occurs, right?
(Also, can something similar be done to avoid ```expect: pass``` in the other function as well?)

Reply:
@atb00ker I think that even better would be something like this:
```
except(SystemExit):
    message = 'Your commit message does not follow our commit message style guidelines'
    self.assertIn(message, captured_output.getvalue()
else: 
    self.fail("Error not raised")
sys.stdout = sys.__stdout__  # reset redirect
```
because we want to reset redirect even if except won't occure
also, I don't understand why we except `Exception` [here](https://github.com/openwisp/openwisp-utils/blob/master/tests/test_project/tests/test_qa.py#L187) if we only except according to function `SystemExit`?

Reply:
> because we want to reset redirect even if except won't occure

Ah, sorry, yep sounds good, if it needs to be executed in anycase it can be `finally:` or outside, both will work! :smile: 
The advantage of putting inside finally is that even if `except` fails for some reason, finally will still execute! :smile:  


> also, I don't understand why we except Exception here if we only except according to function SystemExit?

Even I don't know. Doesn't look good. @ppabcd can you help? :smile: :

Reply:
> Even I don't know. Doesn't look good. @ppabcd can you help? 😄 :

Hmm, when I created this code, I saw the previous unit testing and copy it. After then, I modified to check for a commit message style check because I didn't know how to create a unit testing before.

Reply:
>  I saw the previous unit testing and copy it

@ppabcd Can you send url to the "previous" unit test?

Reply:
@oltarzewskik I think we can skip `Exception`, sometimes wrong code this gets merged. :smile: